rts/lra/lra-coordinator-jar/src/main/java/io/narayana/lra/coordinator/domain/model/LRARecord.java

```diff
@@ -361,18 +375,88 @@ private int tryDoEnd(boolean compensate) {
-                return TwoPhaseOutcome.HEURISTIC_HAZARD;
+                return atEnd(TwoPhaseOutcome.HEURISTIC_HAZARD);
             }
 
             status = ParticipantStatus.Completing; // recovery will figure out the status via the status url
 
-            return TwoPhaseOutcome.FINISH_ERROR;
+            return atEnd(TwoPhaseOutcome.FINISH_ERROR);
         }
 
         updateStatus(compensate);
 
         // if the the request is still in progress (ie accepted is true) let recovery finish it
-        return accepted ? TwoPhaseOutcome.HEURISTIC_HAZARD : TwoPhaseOutcome.FINISH_OK;
+        return atEnd(accepted ? TwoPhaseOutcome.HEURISTIC_HAZARD : TwoPhaseOutcome.FINISH_OK);
+    }
+
+    boolean isFinished() {
+        // nested participants must still be able to compensate even if they are closed
+        if (compensateURI == null) {
+            return afterURI != null;
+        }
+
+        switch (status) {
+            case Completed:
+                /* FALLTHRU */
+            case FailedToComplete:
+                return parentId == null; // completed nested LRAs must remain cancellable
+            case Compensated:
+                /* FALLTHRU */
+            case FailedToCompensate:
+                return true;
+            default:
+                return false;
+        }
+    }
+
+    private boolean afterLRARequest(URI target, Transaction lra, String payload) {
+        try {
+            RequestBuilder builder = new RequestBuilder(target)
+                    .request()
+                    .header(LRA.LRA_HTTP_RECOVERY_HEADER, recoveryURI.toASCIIString());
+
+            if (target.equals(afterURI)) {
+                builder.header(LRA.LRA_HTTP_ENDED_CONTEXT_HEADER, lra.getId().toASCIIString());
+            } else {
+                builder.header(LRA.LRA_HTTP_CONTEXT_HEADER, lra.getId().toASCIIString());
+            }
+
+            builder.async(PARTICIPANT_TIMEOUT, TimeUnit.SECONDS);
+
+            ResponseHolder response = target.equals(forgetURI) ? builder.delete() : builder.put(payload, MediaType.TEXT_PLAIN);
+
+            if (response.getStatus() == 200) {
+                return true;
+            }
+        } catch (WebApplicationException e) {
+            if (LRALogger.logger.isInfoEnabled()) {
+                LRALogger.logger.infof("Could not notify URI at %s (%s)", target, e.getMessage());
+            }
+        }
+
+        return false;
+    }
+
+    private int atEnd(int res) {
+        // Only run the post LRA actions if both the LRA and participant are in an end state
+        // check the participant first since it will have been removed from one of the lists
+        if (!isFinished() || !lra.isFinished()) {
+            if (afterURI != null) {
+                return TwoPhaseOutcome.HEURISTIC_HAZARD;
+            }
+
+            return res;
+        }
+
+        // run post LRA actions
+        if (afterURI == null || afterLRARequest(afterURI, lra, lra.getLRAStatus().name())) {
```

Comment:
@mmusgrov please can you expand on the situation where afterURI can be null at this point and what the value of res could be? 

If afterURI is null we will return TwoPhaseOutcome.FINISH_OK but res might not always be that from what I can see in above calls to atEnd(int) - including at least one call that might pass in HEURISTIC_HAZARD. Maybe in those cases afterURI could never be null? But then what about the one which uses a value from retryGetEndStatus or reportFailure

Reply:
This bit of code will only run if it is time to run the post LRA actions. If those actions succeed then we are finished with this record and we return FINISH_OK which removes this record from the intentions list.

In more detail:

1. If the status passed to the atEnd routine is HEURISTIC_HAZARD then the previous check (line 443) `if (!isFinished() || !lra.isFinished()) {` will have been true so the bit of code you have highlighted will not be reached and we return HH or the res value depending on whether or not afterURI is null.
2. The reportFailure route will always return TwoPhaseOutcome.FINISH_OK (since the spec only requires us to log failures). But do note that we have JBTM-3247 "Failed LRA records are reported but they not kept" which when fixed will keep those records (and the fix for that should take account of your current comment).
3. The other route you mention is retryGetEndStatus then:
   - if this returns HH then it is covered by point 1) above
   - if it returns FE (FINISH_ERROR) then it ought to have been covered by point 2) above where failures should be reported but the log can be removed. Although I do notice in this routine that the failure will not be reported using a logging statement but that is part of the fix for JBTM-3247 (keep failed records).

Reply:
Thanks Mike. Regarding the missing logging you mentioned "will not be reported using a logging statement" - do you mean this particular detection of the LRAStatus being FailedToClose or FailedToCancel was not logged but another one would have been or that it might not have been reported at all? If the latter, it would seem different to https://issues.redhat.com/browse/JBTM-3247

Please can you add the conditions necessary to remember to verify JBTM-3247 text so we remember to check them?

Reply:
I mean not logged at all. I will send all failed outcomes through the reportFailure() route so that they will always be reported on the console. But note that JBTM-3247 keeps the failed records and they are available for inspection using the RecoveryCoordinator resource in the same way that we allow inspection of recovering LRAs. At least that is the way I have resolved JBTM-3247. But my fix for JBTM-3247 requires this current PR so can I have your approval so that I can raise a PR for it.

Reply:
Thanks Mike